### PR TITLE
🐛 tilt: ensure .tiltbuild/bin directory is created early enough, add tilt troubleshooting guide

### LINKED ISSUE #9162: tilt up with podman fails
### What steps did you take and what happened?

tilt with podman is faling

```
$ kind version
kind v0.20.0 go1.20.5 darwin/amd64
```

```
$ DOCKER_HOST=unix:///run/user/$(id -u)/podman/podman.sock tilt up
```

```
$ cat tilt-setting.yaml

default_registry: "gcr.io/you-project-name-here"
enable_providers:
  - kubeadm-bootstrap
  - kubeadm-control-plane
deploy_cert_manager: false
```

tilt error with webhook diabled

```
Loading Tiltfile at: /Users/karthikkn/k-workspace/cluster-api/Tiltfile
local: command -v kubectl || true
local: docker version || podman version
local: go env GOOS
 → darwin
local: go env GOARCH
 → amd64
local: kubectl get cm -n kube-public local-registry-hosting || true
local: make -B tilt-prepare && ./hack/tools/bin/tilt-prepare --tools kustomize,envsubst,clusterctl --tilt-settings-file ./tilt-settings.yaml
 → cd hack/tools; go build -tags=tools -o bin/tilt-prepare sigs.k8s.io/cluster-api/hack/tools/tilt-prepare
 → I0810 15:27:43.923934    5031 main.go:161] [main] started
 → I0810 15:27:43.926355    5031 main.go:423] [tools] task group started
 → I0810 15:27:43.926417    5031 main.go:475] [tools/envsubst] task started
 → I0810 15:27:43.926473    5031 main.go:475] [tools/kustomize] task started
 → I0810 15:27:43.926473    5031 main.go:475] [tools/clusterctl] task started
 → I0810 15:27:44.561459    5031 main.go:477] [tools/envsubst] task completed, elapsed: 634.998748ms
 → I0810 15:27:44.563695    5031 main.go:477] [tools/kustomize] task completed, elapsed: 637.137235ms
 → I0810 15:27:56.487972    5031 main.go:477] [tools/clusterctl] task completed, elapsed: 12.561213311s
 → I0810 15:27:56.488029    5031 main.go:425] [tools] task group completed, elapsed: 12.561394697s
 → I0810 15:27:56.488068    5031 main.go:423] [resources] task group started
 → I0810 15:27:56.488120    5031 main.go:475] [resources/clusterctl.crd] task started
 → I0810 15:27:56.488123    5031 main.go:475] [resources/kubeadm-bootstrap] task started
 → I0810 15:27:56.488727    5031 main.go:475] [resources/kubeadm-control-plane] task started
 → I0810 15:27:56.488739    5031 main.go:475] [resources/core] task started
 → I0810 15:27:56.513867    5031 main.go:751] [resources/clusterctl.crd] no changes in the generated yaml
 → I0810 15:27:56.513899    5031 main.go:477] [resources/clusterctl.crd] task completed, elapsed: 25.761284ms
 → I0810 15:27:57.390745    5031 main.go:751] [resources/kubeadm-bootstrap] no changes in the generated yaml
 → I0810 15:27:57.390761    5031 main.go:477] [resources/kubeadm-bootstrap] task completed, elapsed: 902.603905ms
 → I0810 15:27:57.391334    5031 main.go:751] [resources/kubeadm-control-plane] no changes in the generated yaml
 → I0810 15:27:57.391343    5031 main.go:477] [resources/kubeadm-control-plane] task completed, elapsed: 902.555328ms
 → I0810 15:27:57.499042    5031 main.go:751] [resources/core] no changes in the generated yaml
 → I0810 15:27:57.499059    5031 main.go:477] [resources/core] task completed, elapsed: 1.010263075s
 → I0810 15:27:57.499092    5031 main.go:425] [resources] task group completed, elapsed: 1.010994819s
 → I0810 15:27:57.499101    5031 main.go:187] [main] completed, elapsed: 13.574701223s
local: tee bootstrap/kubeadm/.tiltbuild/bin//Dockerfile
Traceback (most recent call last):
  /Users/karthikkn/k-workspace/cluster-api/Tiltfile:637:17: in <toplevel>
  /Users/karthikkn/k-workspace/cluster-api/Tiltfile:408:24: in enable_providers
  /Users/karthikkn/k-workspace/cluster-api/Tiltfile:348:23: in enable_provider
  /Users/karthikkn/k-workspace/cluster-api/Tiltfile:278:14: in build_docker_image
Error in local: command "tee bootstrap/kubeadm/.tiltbuild/bin//Dockerfile" failed.
error: exit status 1
stdout: "\n# Tilt image\nFROM golang:1.20.7 as tilt-helper\n# Install delve. Note this should be kept in step with the Go release minor version.\nRUN go install github.com/go-delve/delve/cmd/dlv@v1.20\n# Support live reloading with Tilt\nRUN wget --output-document /restart.sh --quiet https://raw.githubusercontent.com/tilt-dev/rerun-process-wrapper/master/restart.sh  &&     wget --output-document /start.sh --quiet https://raw.githubusercontent.com/tilt-dev/rerun-process-wrapper/master/start.sh &&     chmod +x /start.sh && chmod +x /restart.sh && chmod +x /go/bin/dlv &&     touch /process.txt && chmod 0777 /process.txt `# pre-create PID file to allow even non-root users to run the image`\n\n\n\nFROM golang:1.20.7 as tilt\nWORKDIR /\nCOPY --from=tilt-helper /process.txt .\nCOPY --from=tilt-helper /start.sh .\nCOPY --from=tilt-helper /restart.sh .\nCOPY --from=tilt-helper /go/bin/dlv .\nCOPY $binary_name .\n\n"
stderr: "tee: bootstrap/kubeadm/.tiltbuild/bin//Dockerfile: No such file or directory\n"
```

tilt error with webhook enabled in tilt-setting.yaml

```
Loading Tiltfile at: /Users/karthikkn/k-workspace/cluster-api/Tiltfile
local: command -v kubectl || true
local: docker version || podman version
local: go env GOOS
 → darwin
local: go env GOARCH
 → amd64
local: kubectl get cm -n kube-public local-registry-hosting || true
local: make -B tilt-prepare && ./hack/tools/bin/tilt-prepare --tools kustomize,envsubst,clusterctl --tilt-settings-file ./tilt-settings.yaml
 → cd hack/tools; go build -tags=tools -o bin/tilt-prepare sigs.k8s.io/cluster-api/hack/tools/tilt-prepare
 → I0810 15:31:01.138415    5767 main.go:161] [main] started
 → I0810 15:31:01.141678    5767 main.go:423] [tools] task group started
 → I0810 15:31:01.141725    5767 main.go:475] [tools/kustomize] task started
 → I0810 15:31:01.141748    5767 main.go:475] [tools/envsubst] task started
 → I0810 15:31:01.141780    5767 main.go:475] [tools/clusterctl] task started
 → I0810 15:31:01.807255    5767 main.go:477] [tools/kustomize] task completed, elapsed: 665.470656ms
 → I0810 15:31:01.813225    5767 main.go:477] [tools/envsubst] task completed, elapsed: 671.414361ms
 → I0810 15:31:13.714953    5767 main.go:477] [tools/clusterctl] task completed, elapsed: 12.572886089s
 → I0810 15:31:13.715005    5767 main.go:425] [tools] task group completed, elapsed: 12.57305342s
 → I0810 15:31:13.717010    5767 main.go:423] [resources] task group started
 → I0810 15:31:13.717049    5767 main.go:475] [resources/cert-manager-controller] task started
 → I0810 15:31:13.717075    5767 main.go:475] [resources/cert-manager-cainjector] task started
 → I0810 15:31:13.717069    5767 main.go:475] [resources/core] task started
 → I0810 15:31:13.717117    5767 main.go:475] [resources/kubeadm-bootstrap] task started
 → I0810 15:31:13.717125    5767 main.go:475] [resources/cert-manager-webhook] task started
 → I0810 15:31:13.717172    5767 main.go:475] [resources/clusterctl.crd] task started
 → I0810 15:31:13.717161    5767 main.go:475] [resources/cert-manager] task started
 → I0810 15:31:13.717288    5767 main.go:475] [resources/kubeadm-control-plane] task started
 → No default config file available
 → I0810 15:31:13.720482    5767 main.go:477] [resources/cert-manager-webhook] task completed, elapsed: 3.246211ms
 → I0810 15:31:13.721545    5767 main.go:425] [resources] task group completed, elapsed: 4.37391ms
 → E0810 15:31:13.721751    5767 main.go:486] [resources/cert-manager-controller] Recovered from panic: send on closed channel
 → I0810 15:31:13.721806    5767 main.go:477] [resources/cert-manager-controller] task completed, elapsed: 4.747677ms
 → I0810 15:31:13.721817    5767 main.go:477] [resources/cert-manager-cainjector] task completed, elapsed: 4.646655ms
 → F0810 15:31:13.722398    5767 main.go:184] [main] failed to prepare tilt resources: [[resources/cert-manager-webhook] failed to pull quay.io/jetstack/cert-manager-webhook:v1.12.2: failure determining if the image exists in local cache: quay.io/jetstack/cert-manager-webhook:v1.12.2: failure listing container image: quay.io/jetstack/cert-manager-webhook:v1.12.2: Cannot connect to the Docker daemon at unix:///run/user/501/podman/podman.sock. Is the docker daemon running?, [resources/cert-manager-cainjector] failed to pull quay.io/jetstack/cert-manager-cainjector:v1.12.2: failure determining if the image exists in local cache: quay.io/jetstack/cert-manager-cainjector:v1.12.2: failure listing container image: quay.io/jetstack/cert-manager-cainjector:v1.12.2: Cannot connect to the Docker daemon at unix:///run/user/501/podman/podman.sock. Is the docker daemon running?]
Traceback (most recent call last):
  /Users/karthikkn/k-workspace/cluster-api/Tiltfile:631:12: in <toplevel>
  /Users/karthikkn/k-workspace/cluster-api/Tiltfile:487:10: in prepare_all
Error in local: command "make -B tilt-prepare && ./hack/tools/bin/tilt-prepare --tools kustomize,envsubst,clusterctl --tilt-settings-file ./tilt-settings.yaml" failed.
error: exit status 1
```

### What did you expect to happen?

Successfully able to run tilt

### Cluster API version

Latest main

$ git log --oneline
f9942a97b (HEAD -> main, origin/main, origin/HEAD) Merge pull request #9159 from kubernetes-sigs/dependabot/go_modules/hack/tools/google.golang.org/api-0.136.0
f1ef07941 :seedling: Bump google.golang.org/api in /hack/tools
d7229e5cc Merge pull request #9119 from g-gaston/update-versions-1-6

### Kubernetes version

$ kubectl version --short
Flag --short has been deprecated, and will be removed in the future. The --short output will become the default.
Client Version: v1.27.3
Kustomize Version: v5.0.1
Server Version: v1.27.3

### Anything else you would like to add?

_No response_

### Label(s) to be applied

/kind bug
One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels.

### PR DESCRIPTION
**Pull request by bengentil**
**What this PR does / why we need it**:

The `context + "/.tiltbuild/bin/"` directory is created by a `local_resource` 
but `local_resource`s happens after `local` commands, including the one to create the Dockerfile for podman.

This PR ensures that the directory is present before writing the Dockerfile.
`mkdir -p` is idempotent so it's not an issue to call it twice for podman.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #9162

### CODE REVIEW

**Comment by fabriziopandini** [comment]
/ok-to-test
I don't have an environment to test, but the change seems ok
Also, it would be great to have tips&tricks to use podman (e.g https://github.com/kubernetes-sigs/cluster-api/issues/9162#issuecomment-1674699615) to surface in the CAPI book

**Comment by sbueringer** [comment]
I think I hit the same or a similar issue on a MacBook with Apple Silicon and Docker.

In my case I first started tilt and waited until everything was up. Then I added a new provider:
* Docker image build failed
* binary built then worked

I think the problem was that there is no dependency from the Docker image build to the binary build. Usually this is done in the right order because the Docker image build has to wait for the deployment of all CRDs etc. ("uncategorized"). But if this is not the case and the targets run in the wrong order the build will fail.

So I think this issue might not be podman specific but basically it can always occur when a provider is deployed for the first time and the binary was not build before the Docker image build was triggered

**Comment by sbueringer** [comment]
tl;dr. Maybe let's just add this mkdir for both cases and not only to Podman?

**Comment by bengentil** [comment]
> Also, it would be great to have tips&tricks to use podman (e.g https://github.com/kubernetes-sigs/cluster-api/issues/9162#issuecomment-1674699615) to surface in the CAPI book

@fabriziopandini There is a documentation on how to use podman: https://cluster-api.sigs.k8s.io/developer/tilt.html#using-podman
but maybe a small troubleshoot guide or a summary like in the issue could be helpful yes :+1: 

@sbueringer This directory is required  early in `build_docker_image` only for podman as  a Dockerfile is written via a `local` call , docker receive the Dockerfile dynamically via the field `dockerfile_contents` when needed.

From my tests I saw that `local` are being executed immediately but `local_resource` (and other resources like `docker_build`) seems to be executed asynchronously (and re-called  based on the dependencies)

What we could do is  separate the `local_resource` in `build_go_binary` to have:
- 1 `local` to run the `mkdir -p`
- 1 `local_resource` to call the build command

So the directory will be always present before any other resources. This will not completely fix the issue you describe as if the `build_go_binary` is triggered before the build `local_resource` ,  the binaries will be missing.

Please advise, any solution is fine with me :wink:

**Comment by sbueringer** [comment]
If I understand it correctly a local with mkdir would create the directory instantly, so this would be before local_resource and docker_build?

**Comment by bengentil** [comment]
> If I understand it correctly a local with mkdir would create the directory instantly, so this would be before local_resource and docker_build? (and then solve the issue completely)

Yes

**Comment by sbueringer** [comment]
I would go with that

**Comment by bengentil** [comment]
@sbueringer I moved the mkdir to a `local`

@fabriziopandini I pushed a troubleshooting guide to explain how to fix all issues seen in #9162 
let me know if you want me to push that in a separate PR

**Comment by sbueringer** [comment]
Tested it with a MacBook with M2 Pro + Docker, worked as expected

Thank you very much!

Really nice troubleshooting guide!
